src/transform/src/analysis/equivalences.rs

```diff
@@ -554,29 +559,55 @@ impl EquivalenceClasses {
 
     /// Produce the equivalences present in both inputs.
     pub fn union(&self, other: &Self) -> Self {
-        // TODO: seems like this could be extended, with similar concepts to localization:
-        //       We may removed non-shared constraints, but ones that remain could take over
-        //       and substitute in to retain more equivalences.
-
-        // For each pair of equivalence classes, their intersection.
-        let mut equivalences = EquivalenceClasses {
-            classes: Vec::new(),
-            remap: Default::default(),
-        };
-        for class1 in self.classes.iter() {
-            for class2 in other.classes.iter() {
-                let class = class1
-                    .iter()
-                    .filter(|e1| class2.iter().any(|e2| e1 == &e2))
-                    .cloned()
-                    .collect::<Vec<_>>();
-                if class.len() > 1 {
-                    equivalences.classes.push(class);
+        self.union_many(&[other])
+    }
+
+    /// The equivalence classes of terms equivalent in all inputs.
+    ///
+    /// This method currently misses opportunities, because it only looks for exactly matches in expressions,
+    /// which may not include all possible matches. For example, `f(#1) == g(#1)` may exist in one class, but
+    /// in another class where `#0 == #1` it may exist as `f(#0) == g(#0)`.
+    pub fn union_many(&self, others: &[&Self]) -> Self {
+        if others.is_empty() {
+            self.clone()
+        } else {
+            // Map from expression to an equivalence class identifier.
+            // That identifier is a pair of integer and expression reference, where
+            // the integer is a bit of a "hashcons" of accumulated class representatives.
+            let mut intersection: BTreeMap<&MirScalarExpr, usize> = Default::default();
```

Comment:
Yeah, looks like it! I was expecting to need to intersect them, but the code evolved to not need that. Moreover, the `flat_map` is really just a `retain_mut` on a vector, so that a bunch of allocation saved.